aten/src/ATen/native/quantized/cuda/fake_quantize_per_tensor_affine.cu

```diff
@@ -0,0 +1,164 @@
+#include <ATen/ATen.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/core/op_registration/op_registration.h>
+#include <cmath>
+
+/* FakeQuantize Op for PerTensorAffine quantization scheme */
+namespace at { namespace native {
+namespace {
+/* Fake-quantizes the 'inputs' tensor.
+Args:
+  X: Forward input tensor.
+  dY: Backward input tensor (_backward op only).
+  scale: scale of per tensor affine quantization
+  zero_point: zero_point of per tensor affine quantization
+  num_bits: Number of quantization bits.
+  quant_delay: Count of global steps for which to delay the quantization.
+               See note below.
+  iter: The current quantization iteration used for `quant_delay`.
+Returns:
+  Quantized tensor (double dtype).
+
+Notes:
+  - quant_delay might be set to non-zero to help weights stabilize in the
+    beginning of the training.
+  - quantization range [0, 2^bits - 1]
+*/
+
+class FakeQuantizePerTensorAffineOp_forward : public c10::OperatorKernel {
+ public:
+  at::Tensor operator()(
+      at::Tensor X,
+      double scale,
+      int64_t zero_point,
+      int64_t num_bits = 8,
+      int64_t quant_delay = 0,
+      int64_t iter = 0
+    ) {
+    // Sanity checks.
+    AT_CHECK(X.is_cuda());
+    AT_CHECK(X.scalar_type() == ScalarType::Float);
+    if (num_bits > 32 || num_bits < 1) {
+      throw std::invalid_argument("`num_bits` should be in the [1, 32] range.");
+    }
+    if (zero_point < 0) {
+      throw std::invalid_argument("`zero_point` must be a positive integer.");
+    }
+    if (quant_delay < 0) {
+      throw std::invalid_argument("`quant_delay` must be a positive integer.");
+    }
+
+    if (quant_delay != 0 && iter < 0) {
+      throw std::invalid_argument(
+        "`iter` must be >=0 for non-zero `quant_delay`");
+    }
+
+    auto Y = at::empty_like(X);
+
+    if (quant_delay > 0 && iter <= quant_delay) {
+      Y.copy_(X);  // We might want to just return the input here.
+      return Y;
+    }
+
+    float inv_scale = 1.0f / scale;
+    const float quant_min = 0;
+    const float quant_max = (1 << num_bits) - 1;
+    at::cuda::CUDA_tensor_apply2<float, float>(
+        X,
+        Y,
+        [=] __device__ (
+            const float& input_val,
+            float& result_val) {
+          result_val = (fminf(quant_max, fmaxf(quant_min, (std::floor(input_val * inv_scale + 0.5f) + zero_point))) - zero_point) * scale;
```

Comment:
This behavior is not the same as the current implementation of the quantizer: if you use `std::floor(x + 0.5)` while the quantizer implements `std::round(x)`, the behavior at tie breaking will differ. Also, zero_point should be inside the rounding logic (again, because of the rounding tie-breaking).

Reply:
I believe this is trying to implement `nearbyint`, are we using `round` now?

Reply:
changed to `round` and after adding zero_point